net/peer_tracker.go

```diff
@@ -18,32 +22,99 @@ var logPeerTracker = logging.Logger("peer-tracker")
 type PeerTracker struct {
 	// mu protects peers
 	mu sync.RWMutex
-	// peers maps stringified peer.IDs to info about their chains
-	peers map[string]*types.ChainInfo
+	// peers maps peer.IDs to info about their chains
+	peers    map[peer.ID]*types.ChainInfo
+	trusted  map[peer.ID]bool
+	updateFn updatePeerFn
 }
+type updatePeerFn func(ctx context.Context, p peer.ID) (*types.ChainInfo, error)
 
 // NewPeerTracker creates a peer tracker.
-func NewPeerTracker() *PeerTracker {
+func NewPeerTracker(trust ...peer.ID) *PeerTracker {
+	trustedSet := make(map[peer.ID]bool)
+	for _, t := range trust {
+		trustedSet[t] = true
+	}
 	return &PeerTracker{
-		peers: make(map[string]*types.ChainInfo),
+		peers:   make(map[peer.ID]*types.ChainInfo),
+		trusted: trustedSet,
+	}
+}
+
+// SetUpdateFn sets the update function `f` on the peer tracker. This function is a prerequisite
+// to the UpdateTrusted logic.
+func (tracker *PeerTracker) SetUpdateFn(f updatePeerFn) {
+	tracker.updateFn = f
+}
+
+// SelectHead returns the chain info with the greatest height being tracked. If `trust` is true then
+// only the trusted tracked nodes will be considered. An error is returned if no peers are in the tracker.
+func (tracker *PeerTracker) SelectHead(trust bool) (types.ChainInfo, error) {
+	var heads []*types.ChainInfo
+	if trust {
+		heads = tracker.ListTrusted()
+	} else {
+		heads = tracker.List()
+	}
+	if len(heads) == 0 {
+		return types.ChainInfo{}, errors.New("no peers tracked")
 	}
+	sort.Slice(heads, func(i, j int) bool { return heads[i].Height > heads[j].Height })
+	return *heads[0], nil
+}
+
+// UpdateTrusted updates ChainInfo for all trusted peers.
+func (tracker *PeerTracker) UpdateTrusted(ctx context.Context) error {
+	return tracker.updatePeers(ctx, tracker.TrustedPeers()...)
+}
+
+// updatePeers will run the trackers updateFn on each peer in `ps` in parallel, if updating any of
+// the peers in `ps` fails then the first non-nil error is returned.
+func (tracker *PeerTracker) updatePeers(ctx context.Context, ps ...peer.ID) error {
+	if tracker.updateFn == nil {
+		// programmer error
+		panic("canot call PeerTracker peer update logic without setting an update function")
+	}
+	grp, ctx := errgroup.WithContext(ctx)
+	for _, p := range ps {
+		peer := p
+		grp.Go(func() error {
```

Comment:
I expect updating the PeerTrackers to require going to the network, because of this I decided to perform this operation concurrently.

Reply:
Sounds good to me, currently `SelectHead` will return an error if at least one update fails, I will modify s.t. it only errors if all fail.